### PR TITLE
[ARM] az group export: Add new parameters `--skip-resource-name-params` and `--skip-all-params` to support skip parameterization

### PR DESCRIPTION
**Pull request by jorgecotillo**
**Description**  

- Add --skip-resource-name-params argument which removes the use of parameters from name properties
- Add --skip-all-params argument which removes the use of all parameters from the response
- Add 

--resource-ids argument to filter the results based on a string containing whitespace-separated 

**Testing Guide**  
- To export all resource templates from a resource group without resource name parameterization: **az group export --name contoso_rg --skip-resource-name-params --resource-ids "*"**
- To export all resource templates from a resource group without parameters: **az group export --name contoso_rg --skip-all-params --resource-ids "*"**

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
add to S170

**Comment by anthony-c-martin** [comment]
LGTM!

**Comment by zhoxing-ms** [comment]
I suggest that it might be better to change the PR title, like: 
```
[ARM] az resource group export: Add new parameter `--skip-resource-name-parameterization` and `--skip-all-parameterization` to support xxx
```

PR title writing guide: [link](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests)

**Comment by zhoxing-ms** [comment]
Testing Guide
> To export all resource templates from a resource group without resource name parameterization: az resource group export --name contoso_rg --skip-resource-name-parameterization --resource-ids "\*"
> To export all resource templates from a resource group without parameters: az resource group export --name contoso_rg --skip-all-parameterization --resource-ids "\*"

This command should actually be `az group export`, right?

File: .vscode/launch.json [hunk 1]
```diff
@@ -5,7 +5,7 @@
             "name": "Azure CLI Debug (Integrated Console)",
             "type": "python",
             "request": "launch",
-            "pythonPath": "${config:python.pythonPath}",
+            "pythonPath": "${config:python.interpreterPath}",
```
Comment (zhoxing-ms) [inline .vscode/launch.json:8]: Why is it modified here?

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 2]
```diff
@@ -1025,11 +1025,14 @@ def update_resource_group(instance, tags=None):
 
 
 def export_group_as_template(
-        cmd, resource_group_name, include_comments=False, include_parameter_default_value=False):
+        cmd, resource_group_name, resource_ids=None, include_comments=False, include_parameter_default_value=False, skip_resource_name_parameterization=False, skip_all_parameterization=False):
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:1028]: `resource_ids=None` It is recommended to put the new parameters after the existing parameter of the method

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 3]
```diff
@@ -1038,10 +1041,24 @@ def export_group_as_template(
...
+
+    resources = []
+    if resource_ids is None or resource_ids == "*":
+        resources = ["*"]
+    else:
+        for i in resource_ids.split():
+            if is_valid_resource_id(i):
+                resources.append(i)
+            else:
+                raise CLIError('az resource: error: argument --resource_ids: invalid ResourceId value: \'%s\'' % i)
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:1058]: `--resource_ids` --> `--resource-ids`

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py [hunk 4]
```diff
@@ -55,6 +55,32 @@ def test_resource_group(self, resource_group):
...
+    @ResourceGroupPreparer(name_prefix='cli_test_rg_scenario')
+    def test_resource_group_export_skip_all_parameterization(self, resource_group):
+
+        self.kwargs.update({
+            'vnet': 'vnet1'
+        })
+
+        self.cmd('network vnet create -g {rg} -n {vnet}')
+        self.kwargs['vnet_id'] = self.cmd('network vnet show -g {rg} -n {vnet}').get_output_in_json()['id']
+        result = self.cmd('group export --name {rg} --resource-ids "{vnet_id}" --skip-all-parameterization --query "contentVersion"')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py:67]: Could you add checks to this test to confirm whether the `--skip-all-parameterization` is effective?

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py [hunk 5]
```diff
@@ -55,6 +55,32 @@ def test_resource_group(self, resource_group):
...
+    @ResourceGroupPreparer(name_prefix='cli_test_rg_scenario')
+    def test_resource_group_export_skip_resource_name_parameterization(self, resource_group):
+
+        self.kwargs.update({
+            'vnet': 'vnet1'
+        })
+
+        self.cmd('network vnet create -g {rg} -n {vnet}')
+        self.kwargs['vnet_id'] = self.cmd('network vnet show -g {rg} -n {vnet}').get_output_in_json()['id']
+        result = self.cmd('group export --name {rg} --resource-ids "{vnet_id}" --skip-resource-name-parameterization --query "contentVersion"')
```
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py:80]: Could you add checks to this test to confirm whether the `--skip-resource-name-parameterization` is effective?

File: src/azure-cli/azure/cli/command_modules/resource/custom.py (hunk 2 above)
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py:1028]: `--skip-all-parameterization` and `--skip-resource-name-parameterization` are too long for the customers to input. Could you make them shorter?

Comment (apclouds) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py:1028]: How about --skip-all-params and --skip-resource-name-params?

**Comment by jorgecotillo** [comment]
Can someone give me a hint on how to solve the following issue:

AssertionError: No match for the request (<Request (PUT) https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/cli_test_rg_scenario000001/providers/Microsoft.Network/virtualNetworks/vnet1?api-version=2020-04-01>) was found. Can't overwrite existing cassette ('/home/vsts/work/1/s/src/azure-cli/azure/cli/command_modules/resource/tests/latest/recordings/test_resource_group_export_skip_all_params.yaml') in your current record mode ('once').

When I run the test locally, I get vnet api-version 2020-03-01, I just merged the latest from master and still seeing the same api-version and an additional error: test_resource_scenario is also complaining.

Can I get some assistance please?

**Comment by zhoxing-ms** [comment]
> Can someone give me a hint on how to solve the following issue:
> AssertionError: No match for the request (<Request (PUT) https://management.azure.com/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/cli_test_rg_scenario000001/providers/Microsoft.Network/virtualNetworks/vnet1?api-version=2020-04-01>) was found. Can't overwrite existing cassette ('/home/vsts/work/1/s/src/azure-cli/azure/cli/command_modules/resource/tests/latest/recordings/test_resource_group_export_skip_all_params.yaml') in your current record mode ('once').
> When I run the test locally, I get vnet api-version 2020-03-01, I just merged the latest from master and still seeing the same api-version and an additional error: test_resource_scenario is also complaining.
> Can I get some assistance please?

@jorgecotillo Hi, this problem is because the version of `ResourceType.MGMT_NETWORK` was upgraded in the **dev** branch at 05-21. You need to merge with **dev** branch and retest locally for record.

**Comment by jorgecotillo** [comment]
@zhoxing-ms erged from dev and now seeing different errors ...

**Comment by zhoxing-ms** [comment]
This is a known issue: the problem is caused by an incompatibility with Python 3.8 in the `azure-python-devtools` project on which we were dependent
```bash
self = <contextlib.ExitStack object at 0x7f4b609ab070>, exc_details = ()
    def __exit__(self, *exc_details):
>       received_exc = exc_details[0] is not None
E       IndexError: tuple index out of range
```
We already have a PR follow-up to solve this problem: #13661

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 6]
```diff
@@ -1025,11 +1025,14 @@ def update_resource_group(instance, tags=None):
 
 
 def export_group_as_template(
-        cmd, resource_group_name, include_comments=False, include_parameter_default_value=False):
+        cmd, resource_group_name, include_comments=False, include_parameter_default_value=False, resource_ids=None, skip_resource_name_params=False, skip_all_params=False):
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:1028]: Can we register all new arguments in _params.py? Especially for `--resource-ids`, which should be an array, right?

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 7]
```diff
@@ -1025,11 +1025,14 @@ def update_resource_group(instance, tags=None):
...
     """Captures a resource group as a template.
-    :param str resource_group_name:the name of the resoruce group.
-    :param bool include_comments:export template with comments.
+    :param str resource_group_name: the name of the resource group.
+    :param str resource_ids: a string containing space-separated resource ids to filter the export by. To export all resources, do not specify this argument or supply "*".
+    :param bool include_comments: export template with comments.
     :param bool include_parameter_default_value: export template parameter with default value.
+    :param bool skip_resource_name_params: export template and skip resource name parameterization.
+    :param bool skip_all_params: export template parameter and skip all parameterization.
```
Comment (Juliehzl) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:1035]: For these parameters, I will recommend you to add them as `storage_true` and give short name for them.

File: src/azure-cli/azure/cli/command_modules/resource/custom.py (hunk 6 above)
Comment (jorgecotillo) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py:1028]: @Juliehzl thanks, that's what I was looking for. I noticed in the code that we have an "ids" argument and you can pass the values like:

--ids foo bar baz 

and the code will receive three values, I tried to copy the same, but when I pass multiple resource-ids in my case, I get an error saying that "foo" is not a valid argument, instead if I pass:

--resource-ids "foo bar baz" and split the string, then I get it working.

I wasn't sure what else I needed to add, could you please help me here?

File: src/azure-cli/azure/cli/command_modules/resource/custom.py (hunk 7 above)
Comment (jorgecotillo) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py:1035]: Storage_true is to set those values as true by default right? By our default is false, should it be storage_false instead? I may be off here.

Comment (jorgecotillo) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py:1042]: done.

**Comment by yonzhan** [comment]
add to S171

**Comment by jorgecotillo** [comment]
@Juliehzl can you please confirm the recent updates? Let me know if they look ok to you or if additional updates are needed.

**Comment by jorgecotillo** [comment]
@qianwens please let me know if the PR looks good or if updates are required.

**Comment by apclouds** [comment]
@qianwens please let us know if the PR is good to go or if there are any other updates we need to make. Thanks!

File: src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py [hunk 8]
```diff
@@ -55,6 +55,32 @@ def test_resource_group(self, resource_group):
 
         self.assertEqual('"1.0.0.0"\n', result.output)
 
+    @ResourceGroupPreparer(name_prefix='cli_test_rg_scenario')
+    def test_resource_group_export_skip_all_params(self, resource_group):
```
Comment (qianwens) [inline src/azure-cli/azure/cli/command_modules/resource/tests/latest/test_resource.py:59]: >test_resource_group_export_skip_all_params [](start = 8, length = 42)

please also add test case for parameter --resource-ids

File: src/azure-cli/azure/cli/command_modules/resource/_params.py [hunk 9]
```diff
@@ -365,6 +365,9 @@ def load_arguments(self, _):
     with self.argument_context('group export') as c:
         c.argument('include_comments', action='store_true')
         c.argument('include_parameter_default_value', action='store_true')
+        c.argument('skip_resource_name_params', action='store_true')
```
Comment (qianwens) [inline src/azure-cli/azure/cli/command_modules/resource/_params.py:368]: please help add help for these parameters

File: src/azure-cli/azure/cli/command_modules/resource/custom.py [hunk 10]
```diff
@@ -1045,10 +1048,24 @@ def export_group_as_template(
         export_options.append('IncludeComments')
     if include_parameter_default_value:
         export_options.append('IncludeParameterDefaultValue')
+    if skip_resource_name_params:
+        export_options.append('SkipResourceNameParameterization')
+    if skip_all_params:
+        export_options.append('SkipAllParameterization')
+
+    resources = []
+    if resource_ids is None or resource_ids == "*":
```
Comment (qianwens) [inline src/azure-cli/azure/cli/command_modules/resource/custom.py:1057]: >resource_ids [](start = 31, length = 12)

just to confirm here if it should be resource_ids[0] == "*"?

Comment (jorgecotillo) [reply src/azure-cli/azure/cli/command_modules/resource/custom.py]: updated the code.

**Comment by jorgecotillo** [comment]
@Juliehzl @qianwens @zhoxing-ms if the PR looks good, is it possible to merge it please? Otherwise, let me know if there are additional updates required.

**Pull request merged**